private/protocol/json/jsonutil/unmarshal.go

```diff
@@ -211,3 +227,33 @@ func unmarshalScalar(value reflect.Value, data interface{}, tag reflect.StructTa
 	}
 	return nil
 }
+
+func timeFromUnixString(d string) (time.Time, error) {
+	var secs, nsecs int64
+	if dot := strings.Index(d, "."); dot != -1 {
+		var err error
+		secs, err = strconv.ParseInt(d[:dot], 10, 64)
```

Comment:
Nice part about `Modf` is that it preserves the sign of the value for both whole and decimal components returned. So adding the two components together should equal the original value.

For precision we really have to primary options, 1.)  add the float precision error into the resulting nano seconds. 2.) pick a predefined maximum precision and truncate decimal component to it. e.g. milliseconds.

Option 2 probably is the best bet overall to prevent adding complexity of switching based on exponent usage.

```go
func timeFromUnixFloat(f float64) (time.Time, error) {
	if math.IsInf(f, 0) || math.IsNaN(f) {
		return time.Time{}, fmt.Errorf("invalid number value for Unix time, %v", f)
	}

	w, f := math.Modf(f)
	n := int64(f * 1e3)

	return time.Unix(int64(w), n*1e6), nil
}
```